pkg/policy/mapstate.go

```diff
@@ -255,6 +310,11 @@ func (e *MapStateEntry) MergeReferences(entry *MapStateEntry) {
 	for k := range entry.dependents {
 		e.AddDependent(k)
 	}
+
+	// merge DerivedFromRules
+	if len(entry.DerivedFromRules) > 0 {
+		e.DerivedFromRules.MergeSorted(entry.DerivedFromRules)
```

Comment:
This merge should be a no-op when the entries are deeply equal, and the old value was stored for reverting by the calling function if the entries are not deeply equal.  It would be clearer to make this explicit by not calling into this merge code at all in the case of the entires being deeply equal. I'll add a commit to that effect.